WIP: Improve OLM Developer Documentation

awgreene opened the pull request:
**Description of the change:**
Introduce additional documentation to enable new contributors to the Operator

**Motivation for the change:**
It is the goal of the Operator Framework community to enable new developers to contribute to the toolkit.

**Planned Changes**
- [x] Update CONTRIBUTING.md to include new communication channels
- [ ] Review DEVELOPERS.md documentation and remove/update any stale information
- [ ] Review existing developer documentation and remove/update any stale information
- [x] Add a document that lists related projects that enable OLM to install and manage the lifecycle of operators

**Reviewer Checklist**
- [ ] Implementation matches the proposed design, or proposal is updated to match implementation
- [ ] Sufficient unit test coverage 
- [ ] Sufficient end-to-end test coverage
- [ ] Docs updated or added to `/docs` 
- [ ] Commit messages sensible and descriptive

jchunkins commented on CONTRIBUTING.md:
```diff
@@ -2,18 +2,42 @@
 ## Getting started
 
-- Fork the repository on GitHub
+- Fork and clone the repository on GitHub
```
I think it would be helpful to describe this in more detail. Since the project is go based, your fork should be in the right directory structure and not in your personal GitHub directory structure.
```suggestion
- Fork and clone the repository on GitHub
  1. Fork the operator-lifecycle-manager repo (see [GitHub doc](https://docs.github.com/en/github/getting-started-with-github/fork-a-repo))
  1. Clone your personal GitHub fork (prefer using SSH rather than HTTPS)

     ```bash
     mkdir -p $GOPATH/src/github.com/operator-framework
     cd $GOPATH/src/github.com/operator-framework
     git clone git@github.com:<YOUR PERSONAL ORG>/operator-lifecycle-manager.git
     cd operator-lifecycle-manager
     ```
   1. Add the upstream repo so you can keep your fork in sync
     
      ```bash
      git remote add upstream git@github.com:operator-framework/operator-lifecycle-manager.git
      ```

   1. Eventually you'll want to keep your master branch in sync with the upstream repo using commands like `git rebase upstream/master`
```

benluddy replied:
Users shouldn't need their clone to be under GOPATH anymore (we explicitly set `GO111MODULE=on` in the Makefile), but it'd be worthwhile to tell users to install [at least Go 1.11](https://blog.golang.org/modules2019) for module support.

awgreene replied:
I'll call this out in the next update.

awgreene replied:
Addressed.

jchunkins replied:
See https://github.com/operator-framework/operator-lifecycle-manager/pull/1978#issuecomment-770093499

jchunkins commented on CONTRIBUTING.md:
```diff
@@ -2,18 +2,42 @@
+
+To make the bug report accurate and easy to understand, please try to create bug reports that are:
+
+- Specific. Include as much details as possible: which version, what environment, what configuration, etc.
```
```suggestion
- Specific. Include as much detail as possible: which version, what environment, what configuration, etc.
```

awgreene replied:
The suggested change has been accepted in the latest version of the pull request.

jchunkins commented on CONTRIBUTING.md:
```diff
@@ -24,7 +48,8 @@ This is a rough outline of what a contributor's workflow looks like:
 - Make sure commit messages are in the proper format (see below).
```
This mentions formatting but does not mention anything "below"

awgreene replied:
AFAIK we don't really have hard formatting requirements so I'm just going to remove this bit.

jchunkins commented on CONTRIBUTING.md:
```diff
@@ -24,7 +48,8 @@ This is a rough outline of what a contributor's workflow looks like:
 - Push changes in a topic branch to a personal fork of the repository.
 - Submit a pull request to operator-framework/operator-lifecycle-manager.
-- The PR must receive a LGTM from two maintainers found in the MAINTAINERS file.
+- The PR must receive a LGTM from a reviewer found in the OWNERS file.
```
It appears that a new PR automatically suggests reviewers, but those reviewers don't seem to be obligated to provide a review. If someone provides a review, but does not provide a LGTM, then the issue would sit there forever. How would one ensure that a LGTM eventually gets provided so you can move on to the approval step?

jchunkins replied:
Also, what should we do when we make the suggested changes in a review? Should we click the "re-request review" button for the person that provided the review?

jchunkins replied:
In the section. `- Push changes in a topic branch to a personal fork of the repository.`, I also think it should be made clear the DCO requirement means you should use the --signoff flag when doing commits. If you don't do this then you have to go back and rebase your commits to get the signoff comments in place, which is a hassle.... the process to  do that is:

```bash
git rebase -i ID_OF_COMMIT_BEFORE_FIX
# change pick to edit for the commit(s) needing a signoff and save.
git commit --amend --no-edit --signoff
git rebase --continue
```

jchunkins replied:
Also I followed the guide here https://github.com/kubernetes/community/blob/master/contributors/guide/owners.md#the-code-review-process but this process also did not answer my questions above

awgreene replied:
@jchunkins great questions!

>It appears that a new PR automatically suggests reviewers, but those reviewers don't seem to be obligated to provide a review. If someone provides a review, but does not provide a LGTM, then the issue would sit there forever. How would one ensure that a LGTM eventually gets provided so you can move on to the approval step?

There is no requirement for the suggested reviewers to review the PR, typically I would recommend pinging us using one of the mentioned communication channels. I will update the document to capture this step.

> Also, what should we do when we make the suggested changes in a review? Should we click the "re-request review" button for the person that provided the review?

Probably best to ping us again if all the requested changes have been made and the PR hasn't been re-reviewed. Will updated doc to reflect this.

> In the section. - Push changes in a topic branch to a personal fork of the repository., I also think it should be made clear the DCO requirement means you should use the --signoff flag when doing commits...

Great idea, will be added in the next update.

jchunkins commented on CONTRIBUTING.md:
```diff
@@ -24,7 +48,8 @@ This is a rough outline of what a contributor's workflow looks like:
 - Submit a pull request to operator-framework/operator-lifecycle-manager.
-- The PR must receive a LGTM from two maintainers found in the MAINTAINERS file.
+- The PR must receive a LGTM from a reviewer found in the OWNERS file.
+- The PR must receive an APPROVAL from a approver found in the OWNERS file.
```
I asked about this process in my first PR...

https://github.com/operator-framework/operator-lifecycle-manager/pull/1972#issuecomment-767931204

And I got this response:

https://github.com/operator-framework/operator-lifecycle-manager/pull/1972#issuecomment-768044688

So it might be helpful to clarify how approvals are handled.

jchunkins commented on doc/design/related-operator-framework-projects.md:
```diff
@@ -0,0 +1,60 @@
+
+There are a number of projects that enable OLM to install and manage the lifecycle of an operator. Each of these projects have specific goals and their own GitHub repository. Due to the distributed nature of these projects, it can be difficult to know where a change must be made when introducing a new feature or bug fix to OLM.
+
+It is the goal of this document to keep track of the various projects enable OLM to install and manage the lifecycle of operators.
```
```suggestion
It is the goal of this document to keep track of the various projects which enable OLM to install and manage the lifecycle of operators.
```

awgreene replied:
The suggested change has been accepted in the latest version of the pull request

jchunkins commented on doc/design/related-operator-framework-projects.md:
```diff
@@ -0,0 +1,60 @@
+
+When changing an API you should:
+
+1. Fork the [API repo][api] and create a new branch
```
I have not done anything with this repo yet, but I suspect the fork/clone instructions I included in my earlier comment makes sense here too.

awgreene replied:
Good point - adding the note about go modules here as well.

jchunkins replied:
See https://github.com/operator-framework/operator-lifecycle-manager/pull/1978#issuecomment-770093499

jchunkins reviewed (commented):
Leaving some suggestions on what I've seen so far. This is a great overview of the pieces involved.

awgreene commented:
@jchunkins I believe that I've addressed the opened comments, please mark the conversations as resolved if you are happy with the changes.

camilamacedo86 commented on CONTRIBUTING.md:
```diff
@@ -2,29 +2,57 @@
-- Fork the repository on GitHub
+- Fork and clone the repository on GitHub.
+  -  This project uses [Go Modules](https://blog.golang.org/using-go-modules) and does not need to be cloned inside the `$GOPATH` provided that your installed version of GO is greater than or equal to 1.11.
+- Review the design documentation in the [doc/design](./doc/design) directory
```
Will not these docs be moved to https://olm.operatorframework.io/? So, should not the `https://olm.operatorframework.io/` be inked here instead?

@anik120

kevinrizza replied:
I think  `https://olm.operatorframework.io/` should be our user facing documentation, I think that should stay separate from the architecture and design docs that developers for the project should use.

awgreene replied:
I agree with @kevinrizza on this one.

jchunkins commented:
@awgreene Apparently I don't have the right permissions to mark conversations as resolved. I presume that's because I am not a member of the org, but I could be wrong about that.

awgreene commented:
> @awgreene Apparently I don't have the right permissions to mark conversations as resolved. I presume that's because I am not a member of the org, but I could be wrong about that.

@jchunkins got it - can you "like" any comments I made that addressed your concerns so I can resolve the conversation?

jchunkins commented:
@awgreene I think all of them can be resolved... the only one I paused over was not including docs on how to setup the fork and keep it up to date. I guess that's easily discoverable elsewhere. I can see the motivation for keeping it simple... so maybe just keeping the link to https://docs.github.com/en/github/getting-started-with-github/fork-a-repo would be enough? If you don't think it's useful, then by all means leave it off.

timflannagan commented on CONTRIBUTING.md:
```diff
@@ -2,29 +2,57 @@
 
-- Email: [operator-framework][operator_framework]  
+- Slack: [#olm-dev on the Kubernetes Slack](https://kubernetes.slack.com/archives/C0181L6JYQ2)
+- Google Group: [OLM Dev Google Group](https://groups.google.com/u/1/g/operator-framework-olm-dev)
```
It looks like https://groups.google.com/g/operator-framework-olm-dev also works and doesn't contain (presumably) the user (`/u/1/...`) path encoded into the URL.

timflannagan commented on CONTRIBUTING.md:
```diff
@@ -2,29 +2,57 @@
-- Email: [operator-framework][operator_framework]  
+- Slack: [#olm-dev on the Kubernetes Slack](https://kubernetes.slack.com/archives/C0181L6JYQ2)
+- Google Group: [OLM Dev Google Group](https://groups.google.com/u/1/g/operator-framework-olm-dev)
+- Mailing List: [operator-framework-olm-dev@googlegroups.com](operator-framework-olm-dev@googlegroups.com)
```
This link appears to be broken - it looks like it's linking to a local path in the directory. It wasn't immediately obvious to me where I would be able to find this link either.

timflannagan commented on CONTRIBUTING.md:
```diff
@@ -2,29 +2,57 @@
 - Push changes in a topic branch to a personal fork of the repository.
 - Submit a pull request to operator-framework/operator-lifecycle-manager.
-- The PR must receive a LGTM from two maintainers found in the MAINTAINERS file.
+- The pull request must then be reviewed by members of the operator framework team. Please notify the OLM Development team when you create a pull request using one of the links provided under [communication][communication]. As your PR is reviewed, please attempt to keep any design discussions within the pull request. If you have responded to all the messages on your PR and are concerned that it m ...
```
small nit:

```suggestion
- The pull request must then be reviewed by members of the operator framework team. Please notify the OLM Development team when you create a pull request using one of the links provided under [communication][communication]. As your PR is reviewed, please attempt to keep any design discussions within the pull request. If you have responded to all the messages on your PR and are concerned that it might have been forgotten, please ping us once again using one of the links provided under [communication][communication]. Your PR will merge once:
```

timflannagan commented on CONTRIBUTING.md:
```diff
@@ -2,29 +2,57 @@
 - Create a topic branch from where to base the contribution. This is usually master.
-- Make commits of logical units.
-- Make sure commit messages are in the proper format (see below).
+- Make commits of logical units using the `--signoff` flag.
```
Would it make more sense to call out that we require commits to be signed off (enforced through a DCO check) vs. explicitly specifying you need to use the `--signoff` flag?

timflannagan commented:
Quick drive through as I was going through the existing list of issues and saw that this and the DEVELOPING.md documentation also appears to be out-of-date.
